columnflow/tasks/framework/base.py

```diff
@@ -266,6 +266,197 @@ def has_obj(name):
+
+            config_inst = od.Config(
+                id=0,
+                name="my_config",
+                aux={"default_producer": ["my_producer_1", "my_producer_2"]},
+            )
+
+            params = {
+                "config_inst": config_inst,
+                "producer": None,
+            }
+            resolve_param_values(params)  # sets params["producer"] to ("my_producer_1", "my_producer_2")
+
+            params = {
+                "config_inst": config_inst,
+                "producer": "some_other_producer",
+            }
+            resolve_param_values(params)  # sets params["producer"] to ("some_other_producer",)
+
+        When *default_str* points to a function, it should take *container* and *task_params* as an input.
+
+        Example:
+
+        .. code-block:: python
+            def resolve_param_values(params):
+                params["ml_model"] = ConfigTask.resolve_config_default(
+                    params,
+                    params.get("ml_model"),
+                    container=params["config_inst"]
+                    default_str="default_ml_model",
+                    multiple=True,
+                )
+
+            # A function that chooses the ml_model based on an attibute that is set in an inference_model
+            def default_ml_model(container, **task_params):
+                default_ml_model = None
+
+                # check if task is using an inference model
+                if "inference_model" in task_params.keys():
+                    inference_model = task_params.get("inference_model", None)
+
+                    # if inference model is not set, assume it's the container default
+                    if inference_model in (None, "NO_STR"):
+                        inference_model = container.x.default_inference_model
+
+                    # get the default_ml_model from the inference_model_inst
+                    inference_model_inst = columnflow.inference.InferenceModel._subclasses[inference_model]
+                    default_ml_model = getattr(inference_model_inst, "ml_model_name", default_ml_model)
+
+                    return default_ml_model
+
+                return default_ml_model
+
+            config_inst = od.Config(
+                id=0,
+                name="my_config",
+                aux={"default_ml_model": default_ml_model},
+            )
+
+            @inference_model(ml_model_name="default_ml_model")
+            def my_inference_model(self):
+                # some inference model implementation
+                ...
+
+            params = {"config_inst": config_inst, "ml_model": None, "inference_model": "my_inference_model"}
+            resolve_param_values(params)  # sets params["ml_model"] to "my_ml_model"
+
+            params = {"config_inst": config_inst, "ml_model": "some_ml_model", "inference_model": "my_inference_model"}
+            resolve_param_values(params)  # sets params["ml_model"] to "some_ml_model"
+        """
+
```

Comment:
Can we add an early return here for the case that the param is indeed set to make things faster?

```python
if param not in (None, law.NO_STR, ()):
    return param
```

This would also be a bit more expressive as in "in case the parameter has a value, this function actually does not do anything".

Reply:
sure, but then we'd ignore the `multiple` parameter to cast the param to the correct type

Reply:
You're right. I added a suggestion below.